File: frontend/src/lib/crypto.js
// ─────────────────────────────────────────────────────────
//  u-code — Encryption Utils
//  AES-256-CBC with per-message IV
// ─────────────────────────────────────────────────────────

import crypto from 'crypto';

const ALGORITHM = 'aes-256-cbc';

// Helper: Ensure the secret is exactly 32 bytes (256 bits)
function getSecretKey() {
  const secret = process.env.ENCRYPTION_KEY;
  if (!secret) {
    throw new Error('Missing ENCRYPTION_KEY environment variable');
  }
  // Hash the secret to ensure it's always 32 bytes
  return crypto.createHash('sha256').update(String(secret)).digest();
}

/**
 * Encrypt a plaintext string.
 * Returns { encrypted: hex, iv: hex }
 *
 * @param {string} text
 * @returns {{ encrypted: string, iv: string }}
 */
export function encrypt(text) {
  const iv = crypto.randomBytes(16);
  const key = getSecretKey();
  const cipher = crypto.createCipheriv(ALGORITHM, key, iv);

  let encrypted = cipher.update(text, 'utf8', 'hex');
  encrypted += cipher.final('hex');

  return {
    encrypted,
    iv: iv.toString('hex'),
  };
}

/**
 * Decrypt a hex string using the stored IV.
 * Returns the original plaintext.
 *
 * @param {string} encryptedHex
 * @param {string} ivHex
 * @returns {string}
 */
export function decrypt(encryptedHex, ivHex) {
  const iv = Buffer.from(ivHex, 'hex');
  const key = getSecretKey();
  const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);

  let decrypted = decipher.update(encryptedHex, 'hex', 'utf8');
  decrypted += decipher.final('utf8');

  return decrypted;
}

File: ai_engine/app/services/crypto.py
"""AES-256-CBC encrypt/decrypt — wire-compatible with frontend/src/lib/crypto.js.

Both ai_engine and the frontend must produce/consume the same ciphertext format
so encrypted rows in `supabase_projects` (anon_key_enc, service_key_enc,
db_password_enc + their iv columns) are readable from either service.

Format (matches Node's crypto.createCipheriv exactly):
  - Algorithm: AES-256-CBC, PKCS7 padding (the default for Node's CBC)
  - Key:       SHA-256(ENCRYPTION_KEY env var) — gives 32 bytes deterministically
  - IV:        16 random bytes per message
  - Output:    {"encrypted": <hex>, "iv": <hex>}

Pure stdlib + cryptography. No Supabase / network / async.
"""
from __future__ import annotations

import hashlib
import os
import secrets
from dataclasses import dataclass

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.padding import PKCS7

from app.config import settings

_BLOCK_BYTES = 16  # AES block size


@dataclass(frozen=True)
class Encrypted:
    """Hex-encoded ciphertext + IV. Both fields must be persisted; the IV
    is needed for decryption and is not secret."""
    encrypted: str
    iv: str


def _key() -> bytes:
    """Derive the 32-byte AES key. Frontend uses
    crypto.createHash('sha256').update(secret).digest() — same primitive."""
    secret = settings.ENCRYPTION_KEY or os.environ.get("ENCRYPTION_KEY", "")
    if not secret:
        raise RuntimeError(
            "ENCRYPTION_KEY is not set. ai_engine and the frontend must use "
            "the SAME value so encrypted rows are portable between them."
        )
    return hashlib.sha256(secret.encode("utf-8")).digest()


def encrypt(plaintext: str) -> Encrypted:
    """Encrypt a UTF-8 string. Returns hex-encoded ciphertext + IV."""
    if not isinstance(plaintext, str):
        raise TypeError(f"encrypt() expects str, got {type(plaintext).__name__}")
    iv = secrets.token_bytes(_BLOCK_BYTES)
    cipher = Cipher(algorithms.AES(_key()), modes.CBC(iv)).encryptor()
    padder = PKCS7(_BLOCK_BYTES * 8).padder()
    padded = padder.update(plaintext.encode("utf-8")) + padder.finalize()
    ct = cipher.update(padded) + cipher.finalize()
    return Encrypted(encrypted=ct.hex(), iv=iv.hex())


def decrypt(encrypted_hex: str, iv_hex: str) -> str:
    """Reverse encrypt(). Raises ValueError on malformed inputs."""
    try:
        iv = bytes.fromhex(iv_hex)
        ct = bytes.fromhex(encrypted_hex)
    except ValueError as exc:
        raise ValueError(f"crypto: invalid hex input ({exc})") from exc
    if len(iv) != _BLOCK_BYTES:
        raise ValueError(f"crypto: IV must be {_BLOCK_BYTES} bytes, got {len(iv)}")
    cipher = Cipher(algorithms.AES(_key()), modes.CBC(iv)).decryptor()
    padded = cipher.update(ct) + cipher.finalize()
    unpadder = PKCS7(_BLOCK_BYTES * 8).unpadder()
    plain = unpadder.update(padded) + unpadder.finalize()
    return plain.decode("utf-8")
